### PR TITLE
Recurring payments

### PR DESCRIPTION
**Pull request by ScottDormand96**
https://eaflood.atlassian.net/browse/IWTF-3617

Recurring Payments Job to retrieve the recurring payments due on the current day

### CODE REVIEW

File: packages/sales-api-service/src/schema/transaction.schema.js [hunk 1]
```diff
@@ -106,9 +106,24 @@ const finaliseTransactionRequestSchemaContent = {
...
+      cancelledDate: Joi.string()
+        .isoDate()
+        .optional()
+        .description('Optional field for when recurring payment cancelled')
+        .example(new Date().toISOString()),
+      cancelledReason: buildJoiOptionSetValidator('defra_cancelledreason', 'User Cancelled'),
+      endDate: Joi.string().isoDate().required().description('End of recurring payment').example(new Date().toISOString()),
+      agreementId: Joi.string().required().description('Agreement identification number, Gov.UK Pay field').example(uuidv4()),
+      publicId: Joi.string().required().description('SHA-256 hash of id').example(uuidv4()),
+      status: Joi.number().valid(1).required().description('Status field: 1 for active, 0 for not active'),
```
Comment (jaucourt) [inline packages/sales-api-service/src/schema/transaction.schema.js:124]: This isn't required - it should be set to 1 by default

File: packages/sales-api-service/src/services/recurring-payments.service.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import { Contact, findByDateRange, findByExample, findById, permissionForLicensee, RecurringPayment } from '@defra-fish/dynamics-lib'
+
+export const getRecurringPayments = async date => {
+  // create rp thats active with no cancelled date
+  const recurringPayment = new RecurringPayment()
+  recurringPayment.cancelledDate = null
+  // grab all rp matching no cancelled date and status active
+  const activeRecurringPayments = await findByExample(recurringPayment)
+  // grab all rp matching above + within date range of current date -2,-4,-6,-8,-10
+  const dueRecurringPayments = await findByDateRange(activeRecurringPayments, date)
```
Comment (jaucourt) [inline packages/sales-api-service/src/services/recurring-payments.service.js:10]: See my above comments on the filtering. I think you need to have a method to retrieve the uncancelled, active recurring payments for the specified date

File: packages/sales-api-service/src/services/recurring-payments.service.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      recurringPayment.activePermission,
+      recurringPayment.contact.birthDate,
+      recurringPayment.contact.postcode
+    )
+    recurringPayment.activePermission = permission
+    recurringPaymentsWithPermission = recurringPaymentsWithPermission.concat(recurringPayment)
+  }
+
+  return recurringPaymentsWithPermission
+}
```
Comment (jaucourt) [inline packages/sales-api-service/src/services/recurring-payments.service.js:30]: This isn't the way you should be getting related entities. Have a look at permission.queries.js in dynamics-lib, particularly the expand property of PredefinedQuery.

File: packages/sales-api-service/src/services/transactions/process-transaction-queue.js [hunk 4]
```diff
@@ -81,6 +80,9 @@ export async function processQueue ({ id }) {
 
     entities.push(contact, permission)
 
+    console.log(recurringPayment)
+    console.log(permit.isRecurringPaymentSupported)
```
Comment (jaucourt) [inline packages/sales-api-service/src/services/transactions/process-transaction-queue.js:84]: I don't think you meant to leave these in?

File: packages/sales-api-service/src/services/transactions/process-transaction-queue.js [hunk 5]
```diff
@@ -98,6 +100,8 @@ export async function processQueue ({ id }) {
     }
   }
 
+  console.log('entities: ', entities)
```
Comment (jaucourt) [inline packages/sales-api-service/src/services/transactions/process-transaction-queue.js:103]: Or this...

File: packages/sales-api-service/src/services/transactions/process-transaction-queue.js [hunk 6]
```diff
@@ -142,27 +146,6 @@ const mapToPermission = async (
...
-    recurringPayment = new RecurringPayment()
-    recurringPayment.referenceNumber = transactionRecord.payment.recurring.referenceNumber
-    recurringPayment.mandate = transactionRecord.payment.recurring.mandate
-    recurringPayment.inceptionDay = inceptionMoment.date()
-    recurringPayment.inceptionMonth = inceptionMoment.month()
-    payer = await resolveContactPayload(transactionRecord.payment.recurring.payer)
-    recurringPayment.bindToEntity(RecurringPayment.definition.relationships.payer, payer)
-  }
-  return { recurringPayment, payer }
-}
```
Comment (jaucourt) [inline packages/sales-api-service/src/services/transactions/process-transaction-queue.js:164]: maybe I'm missing something, but as you've removed the processRecurringPayment function, why are you still calling it on line 72?

**Review by jaucourt — CHANGES_REQUESTED** [review]
Great progress here, and much closer to what we're after, but there's still work to do I'm afraid.

File: packages/sales-api-service/src/services/transactions/process-transaction-queue.js (hunk 6 above)
Comment (ScottDormand96) [reply packages/sales-api-service/src/services/transactions/process-transaction-queue.js:164]: I moved to recurring-payments service and called it line 72 instead higher up as had access to contact

File: packages/sales-api-service/src/services/recurring-payments.service.js (hunk 3 above)
Comment (jaucourt) [reply packages/sales-api-service/src/services/recurring-payments.service.js:12]: I don't understand why you're still retrieving Contact and Permission individually; the additional work you've done to retrieve related entities in dynamics-lib should mean you don't have to do this.

File: packages/recurring-payments-job/src/recurring-payments-processor.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { salesApi } from '@defra-fish/connectors-lib'
+
+export const processRecurringPayments = async () => {
+  if (process.env.RUN_RECURRING_PAYMENTS?.toLowerCase() === 'true') {
+    console.log('Recurring Payments job enabled')
```
[the remark this answers is not in the document]
Comment (jaucourt) [reply packages/recurring-payments-job/src/recurring-payments-processor.js:5]: Has this been done?

File: packages/sales-api-service/src/schema/transaction.schema.js [hunk 8]
```diff
@@ -106,9 +106,24 @@ const finaliseTransactionRequestSchemaContent = {
...
     method: buildJoiOptionSetValidator('defra_paymenttype', 'Debit card'),
     recurring: Joi.object({
-      payer: contactRequestSchema,
-      referenceNumber: Joi.string().required().description('The reference number associated with the recurring payment').example(uuidv4()),
-      mandate: Joi.string().required().description('The mandate identifier associated with the recurring payment').example(uuidv4())
+      name: Joi.string().required().description('The default name associated with the recurring payment').example(uuidv4()),
+      nextDueDate: Joi.string()
+        .isoDate()
+        .required()
+        .description('Ten days before the end date of last purchased permission')
```
[the remark this answers is not in the document]
Comment (jaucourt) [reply packages/sales-api-service/src/schema/transaction.schema.js]: This hasn't been changed...

**Review by jaucourt — CHANGES_REQUESTED** [review]
Sorry, still a handful of things to resolve

File: packages/sales-api-service/src/schema/transaction.schema.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ScottDormand96) [reply packages/sales-api-service/src/schema/transaction.schema.js]: it should have updated too:
      nextDueDate: Joi.string()
        .isoDate()
        .required()
        .description('The date of payment for a renewed permission')
        .example(new Date().toISOString()),

File: packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  })
+
+  it('console log displays "Recurring Payments job disabled" when env is false', async () => {
+    process.env.RUN_RECURRING_PAYMENTS = 'false'
+    const consoleLogSpy = jest.spyOn(console, 'log').mockImplementation(jest.fn())
+
+    await processRecurringPayments()
+
+    expect(consoleLogSpy).toHaveBeenCalledWith('Recurring Payments job disabled')
+    consoleLogSpy.mockRestore()
```
Comment (irisfaraway) [inline packages/recurring-payments-job/src/__tests__/recurring-payments-processor.spec.js:20]: Are all these `mockRestore()`s necessary? I don't think we do it elsewhere and it seems like we should be clearing mocks before running other tests anyway.

**Review by irisfaraway — COMMENTED** [review]
Just a question about testing

**Review by irisfaraway — APPROVED** [review]

**Pull request merged**